### PR TITLE
[IMP] mass_mailing: restyle the mass mailing form view to full width

### PR DESCRIPTION
**Pull request by Zynton**
RATIONALE

Currently, editing a mail feels a little claustrophobic because of the boxes
within boxes within boxes design. Purpose of this task is to improve
layout of mailing form view when being in Email Marketing, notably displaying
editor in full width.

SPECIFICATIONS

In this task we remove the chatter and the sheet from the form view so that
the mail editor takes up more horizontal space. The chatter is then moved into
one of the notebook's tabs.

Mass mailing form view currently has a scrollbar for the contents of the
editor: one for the form view itself and one for the sidebar. In consequence
we sometimes end up with three scrollbars side by side, which is ugly and
confusing. In this task we ensure the height of the iframe is always the same
as its contents so there is no need for a scrollbar.

Note that SMS Marketing is untouched as it keeps the old layout. Indeed there
is no need of such changes as it does not use the HTML editor. Actions that
are generic (systray, list view, ...) also use the old display..

LINKS

Task-2545445

### CODE REVIEW

File: addons/mass_mailing/static/src/scss/mass_mailing.scss [hunk 1]
```diff
@@ -12,11 +12,12 @@
...
-    .o_mail_add_emoji {
-        float: none;
-        display: inline-block;
-        bottom: 0;
+
+        .o_mail_add_emoji {
+            float: none;
+            display: inline-block;
+            bottom: 0;
+        }
```
Comment (Zinston) [inline addons/mass_mailing/static/src/scss/mass_mailing.scss:20]: Reminder: This is a fix of an earlier commit, move it there.

File: addons/mass_mailing/static/src/scss/mass_mailing.scss [hunk 2]
```diff
@@ -55,15 +63,19 @@
...
 }
 
 .o_mass_mailing_mailing_form {
     .alert:not(.o_invisible_modifier) + .clearfix.position-relative.o_form_sheet {
-        //hides extra space between form and alert messages
+        // Hides extra space between form and alert messages.
         margin-top: -12px;
     }
+    .alert.alert-info:not(.o_invisible_modifier) {
+        margin: auto 0 auto 12px;
```
Comment (Zinston) [inline addons/mass_mailing/static/src/scss/mass_mailing.scss:76]: Can probably be `auto 0` if I remove the line above.

File: addons/mass_mailing/views/mailing_mailing_views.xml [hunk 3]
```diff
@@ -175,7 +177,7 @@
                                             attrs="{
                                                 'required':[('mailing_model_name','=','mailing.list')],
                                                 'readonly': [('state', 'in', ('sending', 'done'))]
-                                        }"/>
+                                            }"/>
```
Comment (Zinston) [inline addons/mass_mailing/views/mailing_mailing_views.xml:180]: Again a fix of an earlier commit to move there.

File: addons/mass_mailing/views/mailing_mailing_views.xml [hunk 4]
```diff
@@ -54,69 +54,50 @@
...
-                        <button name="action_launch" type="object" class="oe_highlight" string="Send"
-                            attrs="{'invisible': [('state', 'in', ('in_queue',  'sending', 'done'))]}"
-                            confirm="This will send the email to all recipients. Do you still want to proceed ?"/>
-                        <button name="action_schedule" type="object" class="btn-secondary" string="Schedule"
-                            attrs="{'invisible': [('state', 'in', ('in_queue',  'sending', 'done'))]}"/>
-                        <button name="action_test" type="object" class="btn-secondary" string="Test"/>
-                        <button name="action_cancel" type="object" attrs="{'invisible': [('state', '!=', 'in_queue')]}" class="btn-secondary" string="Cancel"/>
-                        <button name="action_retry_failed" type="object" attrs="{'invisible': ['|', ('state', '!=', 'done'), ('failed', '=', 0)]}" class="oe_highlight" string="Retry"/>
+                <form string="Mailing" class="o_mass_mailing_mailing_form" js_class="mass_mailing_form">
+                    <div class="o_mass_mailing_sheet">
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/views/mailing_mailing_views.xml:58]: Form view should avoid unnecessary layers and custom styling at all costs. Why do we need this ? Can't we remove it and define css on o_mass_mailing_mailing_form ?

File: addons/mass_mailing/views/mailing_mailing_views.xml [hunk 5]
```diff
@@ -54,69 +54,50 @@
...
+                        <header style="min-height:31px;">
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/views/mailing_mailing_views.xml:59]: Form view should not hold inline styling. A minimal CSS is accepted in css files, especially here as it is quite custom but form view should be as light as possible to ease its edition.

File: addons/mass_mailing/views/mailing_mailing_views.xml [hunk 6]
```diff
@@ -54,69 +54,50 @@
...
-                                <span name="next_departure_text">This mailing is scheduled for </span>
-                                <field name="next_departure" class="oe_inline"/>.
-                            </strong>
-                        </div>
-                    </div>
+                            <div class="alert alert-info text-center"
+                                role="alert"
+                                attrs="{'invisible': ['&amp;','&amp;',('state', '!=', 'in_queue'),('scheduled', '=', 0),('failed', '=', 0)]}">
+                                <div attrs="{'invisible': ['|', ('state', '!=', 'done'), ('failed', '=', 0)]}">
+                                    <button class="btn-link py-0"
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/views/mailing_mailing_views.xml:73]: Linting: Strange, in hereabove div you indent lines, here you align properties. Somehow you will have to choose your styling (we generally go for indentation and avoid properties alignment in standard views). To propagate to other occurrences.

File: addons/mass_mailing/views/mailing_mailing_views.xml [hunk 7]
```diff
@@ -154,13 +135,33 @@
                                 class="oe_stat_button">
                                 <field name="bounced_ratio" string="Bounced" widget="percentpie"/>
                             </button>
+                            <button name="action_view_traces_sent"
+                                icon="fa-paper-plane"
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/views/mailing_mailing_views.xml:139]: Linting: Globally here some properties could be set on same line to reduce file length.

File: addons/mass_mailing_sms/views/mailing_mailing_views.xml [hunk 8]
```diff
@@ -14,32 +14,30 @@
...
-            <xpath expr="//div[@name='schedule_block']/div[hasclass('row')]" position="inside">
-                <div attrs="{'invisible': ['|', '|', ('mailing_type', '=', 'mail'), ('state', 'in', ['in_queue', 'sending', 'done']), ('schedule_type', '=', 'scheduled')]}">
-                    <label for="sms_force_send" string="Skip Queue"/>
-                    <field name="sms_force_send"/>
-                </div>
-            </xpath>
-            <!-- Headers / Warnings -->
-            <xpath expr="//header" position="after">
+            <form string="Mailing" class="o_mass_mailing_mailing_form">
+                <header style="min-height:31px;">
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_sms/views/mailing_mailing_views.xml:21]: Same here: no custom styling in form views

File: addons/mass_mailing_sms/views/mailing_mailing_views.xml [hunk 9]
```diff
@@ -14,32 +14,30 @@
...
+                    <button name="action_launch" type="object" class="oe_highlight" string="Send"
+                        attrs="{'invisible': ['|', ('mailing_type', '!=', 'mail'), ('state', 'in', ('in_queue', 'sending', 'done'))]}"
+                        confirm="This will send the email to all recipients. Do you still want to proceed ?"/>
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_sms/views/mailing_mailing_views.xml:24]: Still talking about email, which is incorrect in SMS view. I see some other are updated, could you re-check once everything to ensure we don't talk about mail in any sms-based view ? Thanks !

File: addons/mass_mailing_sms/views/mailing_mailing_views.xml [hunk 10]
```diff
@@ -60,113 +58,258 @@
...
-                <page string="SMS Content" name="sms_body" attrs="{'invisible': [('mailing_type', '!=', 'sms')]}">
-                    <field name="body_plaintext" widget="sms_widget"
-                        attrs="{'required': [('mailing_type', '=', 'sms')], 'readonly': [('state', 'in', ('sending', 'done'))]}"
-                        options='{"enable_emojis": True}'/>
-                </page>
-            </xpath>
-            <xpath expr="//field[@name='user_id']" position="after">
-                <field name="sms_allow_unsubscribe" attrs="{'invisible': [('mailing_type', '!=', 'sms')], 'readonly': [('state', 'in', ('sending', 'done'))]}"/>
-            </xpath>
+                <div class="alert alert-info text-center" role="alert" attrs="{'invisible': ['&amp;','&amp;','&amp;','&amp;',('state', '!=', 'in_queue'),('sent', '=', 0),('ignored', '=', 0),('scheduled', '=', 0),('failed', '=', 0)]}">
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_sms/views/mailing_mailing_views.xml:61]: Linting: could be multi-lined to ease reading

File: addons/mass_mailing_sms/views/mailing_mailing_views.xml [hunk 11]
```diff
@@ -60,113 +58,258 @@
...
+                    <div attrs="{'invisible': [('ignored', '=', 0)]}">
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_sms/views/mailing_mailing_views.xml:62]: Why don't you set the invisible directly on button ? It seems the div only adds unnecessary layer. In which case, check to remove other occurences, purpose being to have a layout as flat and as standard (without "div") as possible.

File: addons/mass_mailing_sms/views/mailing_mailing_views.xml [hunk 12]
```diff
@@ -60,113 +58,258 @@
...
-            <xpath expr="//div[@name='contact_ab_pc']" position="attributes">
-                <attribute name="attrs">{'invisible': [('mailing_type', '!=', 'mail')]}</attribute>
-            </xpath>
-            <xpath expr="//label[@for='contact_ab_pc']" position="attributes">
-                <attribute name="attrs">{'invisible': [('mailing_type', '!=', 'mail')]}</attribute>
-            </xpath>
+                <sheet>
+                    <div class="oe_button_box" name="button_box">
+                        <button name="action_view_delivered"
+                            id="button_view_delivered"
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_sms/views/mailing_mailing_views.xml:118]: Same here, some attributes could be grouped on a line to lessen diff and file size (e.g. generally, name + id and type + class)

File: addons/mass_mailing_sms/views/mailing_mailing_views.xml [hunk 13]
```diff
@@ -60,113 +58,258 @@
...
+                        <button name="action_view_bounced"
+                            type="object"
+                            context="{'search_default_filter_bounced': True}"
+                            attrs="{'invisible': [('state', 'in', ('draft','test'))]}"
+                            class="oe_stat_button">
+                            <field name="bounced_ratio" string="Bounced" widget="percentpie"/>
+                        </button>
+                    </div>
+                    <widget name="web_ribbon" title="Archived" bg_color="bg-danger" attrs="{'invisible': [('active', '=', True)]}"/>
+                    <group class="o_mass_mailing_mailing_group">
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_sms/views/mailing_mailing_views.xml:155]: Do we need group with styling on sms view that does not use the editor ?

File: addons/mass_mailing_sms/views/mailing_mailing_views.xml [hunk 14]
```diff
@@ -60,113 +58,258 @@
...
+                        <field name="active" invisible="1"/>
+                        <field name="mailing_type" widget="radio" options="{'horizontal': true}" invisible="1"
+                            attrs="{'readonly': [('state', '!=', 'draft')]}" force_save="1"/>
+                        <field class="o_text_overflow" name="subject" string="Subject" attrs="{'invisible': [('mailing_type', '!=', 'mail')], 'readonly': [('state', 'in', ('sending', 'done'))], 'required': [('mailing_type', '=', 'mail')]}" widget="char_emojis" placeholder="e.g. New Sale on all T-shirts"
+                            help="For an Email, Subject your Recipients will see in their inbox.
+                                For an SMS Text Message, internal Title of the Message."/>
+                        <field class="o_text_overflow" name="sms_subject" string="Title" placeholder="e.g. Black Friday SMS coupon" attrs="{'invisible': [('mailing_type', '!=', 'sms')], 'readonly': [('state', 'in', ('sending', 'done'))], 'required': [('mailing_type', '=', 'sms')]}"/>
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_sms/views/mailing_mailing_views.xml:162]: Could you ensure you always give "name" first ? That the most important think when reading a view, having <field name="..." /> as other parts of there only if you want to go deeper.

File: addons/mass_mailing_sms/views/mailing_mailing_views.xml [hunk 15]
```diff
@@ -240,6 +409,9 @@
                 'default_mailing_type': 'sms',
                 'mailing_sms': True
         }</field>
+        <field name="view_ids" eval="[
```
Comment (tde-banana-odoo) [inline addons/mass_mailing_sms/views/mailing_mailing_views.xml:412]: If necessary, define real ir.actions.act_window.view as it is more maintainable and easier for inheritance / customization.

**Review by tde-banana-odoo — COMMENTED** [review]
Validation: thanks for your work, result seems interesting !

I had a look at views, some global remarks / questions raised. Once updated / answer ping me so that we have another look :) .

File: addons/mass_mailing/static/src/js/mass_mailing_form.js [hunk 16]
```diff
@@ -1,10 +1,107 @@
...
+            return;
+        }
+        this._autoresizeIframe($iframe);
+        const $iframeDoc = $iframe.contents();
+        const $ref = $iframeDoc.find('#iframe_target');
+        if ($ref[0]) {
+            const isFull = window.top.document.body.classList.contains("o_field_widgetTextHtml_fullscreen");
+            $ref.css({
+                display: isFull ? '' : 'flex',
+                'flex-direction': isFull ? '' : 'column',
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_form.js:53]: Strange mix of quote and un-quoted keys.

File: addons/mass_mailing/static/src/js/mass_mailing_form.js [hunk 17]
```diff
@@ -1,10 +1,107 @@
...
 import viewRegistry from 'web.view_registry';
 import config from 'web.config';
 
+const MassMailingFormController = FormController.extend({
+    /**
+     * @override
+     */
+    init() {
+        this._super(...arguments);
+        this._boundTriggerIframeUpdated = this._triggerIframeUpdated.bind(this);
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_form.js:16]: We usually name event-triggered functions "_onXXX" and put them in their own "// Handlers" section, separated from the "// Private" one.

File: addons/mass_mailing/static/src/js/mass_mailing_form.js [hunk 18]
```diff
@@ -1,10 +1,107 @@
...
+            }
+            $content.trigger('scroll');
+        });
+    },
+    /**
+     * Trigger an 'iframe_updated' event on a bus, passing it the mass mailing
+     * iframe jquery element.
+     */
+    _triggerIframeUpdated() {
+        bus.trigger('iframe_updated', this.$('iframe.wysiwyg_iframe, iframe.o_readonly'));
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_form.js:87]: Could directly call "_initIframeResizer" instead of chaining events, for readability?

File: addons/mass_mailing/static/src/js/mass_mailing_form.js [hunk 19]
```diff
@@ -1,10 +1,107 @@
...
+                'flex-direction': isFull ? '' : 'column',
+            });
+            if (this._resizeObserver) {
+                this._resizeObserver.disconnect();
+            }
+            this._resizeObserver = new ResizeObserver(() => this._autoresizeIframe($iframe));
+            this._resizeObserver.observe($ref[0]);
+        }
+        const $content = this.$('.o_content');
+        $content.on('scroll', () => this._repositionSidebar($iframe));
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_form.js:65]: Does this mean that you register a scroll listener every time the "_initIframeResizer" is called? 
Doesn't that end up creating a lot of unnecessary / redundant ones?

Another approach would be to register a scroll listener at the controller level:

```js
events: _.extend({}, FormController.prototype.events, {
    'scroll .o_content': '_onContentScroll',
}
```

And in the "_initIframeResizer", you would update a "this.$iframe = $iframe" reference, to be able to access it in the listener.
(Just some suggestion, maybe it wouldn't work, I don't know the code enough).

File: addons/mass_mailing/static/src/js/mass_mailing_form.js [hunk 20]
```diff
@@ -1,10 +1,107 @@
...
+     * Resize the given iframe so its height fits its contents.
+     *
+     * @param {JQuery} $iframe
+     */
+    _autoresizeIframe($iframe) {
+        const minHeight = $(window).height() - Math.abs($iframe.offset().top) - 6;
+        const $iframeDoc = $iframe.contents();
+        const $themeSelectorNew = $iframeDoc.find('.o_mail_theme_selector_new');
+        if ($themeSelectorNew.length) {
+            const scrollEl = $themeSelectorNew.length ? $themeSelectorNew[0] : this.wysiwyg.$editable[0];
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_form.js:99]: Mmh, seems like you're checking `$themeSelectorNew.length` multiple time, a bit strange.

File: addons/mass_mailing/static/src/js/mass_mailing_form.js [hunk 21]
```diff
@@ -1,10 +1,107 @@
...
+    _triggerIframeUpdated() {
+        bus.trigger('iframe_updated', this.$('iframe.wysiwyg_iframe, iframe.o_readonly'));
+    },
+    /**
+     * Resize the given iframe so its height fits its contents.
+     *
+     * @param {JQuery} $iframe
+     */
+    _autoresizeIframe($iframe) {
+        const minHeight = $(window).height() - Math.abs($iframe.offset().top) - 6;
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_form.js:95]: Lots of magic numbers (6, 12 here below), makes the code a bit difficult to understand, would be nice to add small comments to what they are referring to.

File: addons/mass_mailing/static/src/js/mass_mailing_widget.js [hunk 22]
```diff
@@ -299,6 +299,7 @@ var MassMailingFieldHtml = FieldHtml.extend({
             }
         }
         this.wysiwyg.trigger('reload_snippet_dropzones');
+        core.bus.trigger('iframe_updated', this.wysiwyg.$iframe);
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_widget.js:302]: I'm wondering, since we are in a widget and want to notify our form view, wouldn't a "good old" trigger_up do the trick, instead of using the bus?

File: addons/mass_mailing/static/src/js/mass_mailing_widget.js [hunk 23]
```diff
@@ -373,8 +379,15 @@ var MassMailingFieldHtml = FieldHtml.extend({
...
-        const $codeview = this.$content.parent().find('textarea.o_codeview');
-        $snippetsSideBar.on('click', '.o_codeview_btn', () => this._toggleCodeView($codeview));
+        const $codeview = this.$content.parent().parent().find('textarea.o_codeview');
+        $snippetsSideBar.on('click', '.o_codeview_btn', () => {
+            this._toggleCodeView($codeview);
+            const isFullscreen = !!$(window.top.document).find('.o_mass_mailing_mailing_form_nosheet')[0];
+            $codeview.css('height', isFullscreen ? $(window).height() : '');
+            if ($codeview.hasClass('d-none')) {
+                core.bus.trigger('iframe_updated', this.wysiwyg.$iframe);
+            }
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_widget.js:389]: Can't this be part of a "_toggleCodeView" method override ?

File: addons/web_editor/static/src/js/backend/field_html.js [hunk 24]
```diff
@@ -309,6 +309,7 @@ var FieldHtml = basic_fields.DebouncedField.extend(TranslatableFieldMixin, {
                 }
                 self.$content = $('#iframe_target', self.$iframe[0].contentWindow.document.body);
                 resolver();
+                core.bus.trigger('iframe_updated', self.$iframe);
```
Comment (awa-odoo) [inline addons/web_editor/static/src/js/backend/field_html.js:312]: Could also be in a mass_mailing "_toggleCodeView" override since the iframe_updated thing is only used in mass_mailing anyway.

File: addons/web_editor/static/src/scss/wysiwyg_iframe.scss [hunk 25]
```diff
@@ -15,7 +15,6 @@ iframe.wysiwyg_iframe.o_fullscreen {
 
 body.o_in_iframe {
     background-color: $o-view-background-color;
-    display: flex;
```
Comment (awa-odoo) [inline addons/web_editor/static/src/scss/wysiwyg_iframe.scss:18]: Seems a bit too global, are we sure it's ok to remove that for all usages?

File: addons/mass_mailing/static/src/js/mass_mailing_form.js [hunk 26]
```diff
@@ -100,6 +105,32 @@ const MassMailingFormController = FormController.extend({
             }
         }
     },
+    /**
+     * Reposition the sidebar.
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_form.js:109]: That's a bit light.
Could elaborate based on your commit message.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg_iframe.js [hunk 27]
```diff
@@ -123,9 +123,9 @@ Wysiwyg.include({
                 const $codeview = $('<textarea class="o_codeview d-none"/>');
                 self.$editable.addClass('o_editable oe_structure');
 
+                $iframeTarget.append($codeview);
```
Comment (awa-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg_iframe.js:126]: Is that change directly related to our mass mailing customisations?

**Review by awa-odoo — COMMENTED** [review]
Hello @Zinston 

Since Thibault is still on vacation, Luc asked me if I could validate this since this task is important for the v15.
I will probably have to re-read the code a second time (especially the scroll stuff), but here are already some questions / comments / suggestions.

Thank you for your work!

File: addons/mass_mailing/static/src/js/mass_mailing_widget.js (hunk 23 above)
Comment (Zinston) [reply addons/mass_mailing/static/src/js/mass_mailing_widget.js]: Yup.

File: addons/web_editor/static/src/js/backend/field_html.js (hunk 24 above)
Comment (Zinston) [reply addons/web_editor/static/src/js/backend/field_html.js]: Right now that's the only place we use it but if we're going to trigger an event to notify whenever the iframe is updated, we might as well do it every time the iframe is updated, for future proofinessitude.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg_iframe.js (hunk 27 above)
Comment (Zinston) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg_iframe.js:126]: Yes. A bit further down I move the `#iframe_target` element so it's at the same place no matter the context and this impacts the display of the code view.

File: addons/web_editor/static/src/scss/wysiwyg_iframe.scss (hunk 25 above)
Comment (Zinston) [reply addons/web_editor/static/src/scss/wysiwyg_iframe.scss:18]: I adapted the code where I needed to and didn't find any place where that seemed to cause issues.

**Review by Zynton — COMMENTED** [review]
@awa-odoo Thanks for the review! I addressed your comments. Note there's a new temporary commit appended to this PR which refactors the form controller a little. If you approve it I'll squash it with the rest.

File: addons/mass_mailing/static/src/js/mass_mailing_full_width_form.js [hunk 28]
```diff
@@ -0,0 +1,208 @@
...
+                this.$iframe.height(Math.max(ref.scrollHeight + VERTICAL_OFFSET, minHeight));
+            }
+        }
+    },
+    /**
+     * Reposition the sidebar so it always occupies the full available height.
+     *
+     * @private
+     */
+    _repositionSidebar() {
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_full_width_form.js:60]: Same here, "_repositionTemplateEditorSidebar" ?

File: addons/mass_mailing/static/src/js/mass_mailing_full_width_form.js [hunk 29]
```diff
@@ -0,0 +1,208 @@
...
+            this.$iframe.height(Math.max($themeSelectorNew[0].scrollHeight + VERTICAL_OFFSET, minHeight));
+        } else {
+            const ref = $iframeDoc.find('#iframe_target')[0];
+            if (ref) {
+                this.$iframe.height(Math.max(ref.scrollHeight + VERTICAL_OFFSET, minHeight));
+            }
+        }
+    },
+    /**
+     * Reposition the sidebar so it always occupies the full available height.
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_full_width_form.js:56]: From what I understand, it's more about moving it to follow the scroll position, or is that incorrect?

File: addons/mass_mailing/static/src/js/mass_mailing_full_width_form.js [hunk 30]
```diff
@@ -0,0 +1,208 @@
...
+        bus.off('DOM_updated', this, this._boundOnDomUpdated);
+        this._super(...arguments);
+    },
+
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * Resize the iframe so its height fits its contents.
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_full_width_form.js:36]: Could be nice to add a few use case examples where this is called.

```
Typically called when:
  - A template is picked
  - The enclosing mass mailing form element is resized
  - ...
```

File: addons/mass_mailing/static/src/js/mass_mailing_full_width_form.js [hunk 31]
```diff
@@ -0,0 +1,208 @@
...
+            const isFullscreen = this._isFullScreen();
+            iframeTarget.css({
+                display: isFullscreen ? '' : 'flex',
+                'flex-direction': isFullscreen ? '' : 'column',
+            });
+            if (this._resizeObserver) {
+                this._resizeObserver.disconnect();
+            }
+            this._resizeObserver = new ResizeObserver(this._onResizeIframeContents.bind(this));
+            this._resizeObserver.observe(iframeTarget[0]);
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_full_width_form.js:136]: Isn't that also something you need to register only when your iframe reference changes, in your test here above?

File: addons/mass_mailing/static/src/js/mass_mailing_widget.js [hunk 32]
```diff
@@ -372,7 +392,7 @@ var MassMailingFieldHtml = FieldHtml.extend({
         if (!odoo.debug) {
             $snippetsSideBar.find('.o_codeview_btn').hide();
         }
-        const $codeview = this.$content.parent().find('textarea.o_codeview');
+        const $codeview = this.$content.parent().parent().find('textarea.o_codeview');
```
Comment (awa-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_widget.js:395]: Aren't you better of just calling something like this.$el.find() instead of trying to guess how many times you need to move up the hierarchy before calling find?

**Review by awa-odoo — COMMENTED** [review]
Hello @Zinston 

I like the refactoring of the Controller yes, code is better organized and more readable, so you can keep it.
I added a few picky remarks but I think we are getting close here.

(Side note: the scrollbar stuff still makes my head hurt a little bit ^^)

Thanks!

File: addons/mass_mailing/static/src/js/mass_mailing_form.js (hunk 16 above)
Comment (awa-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_form.js]: No it's ok, keep it like that then.

**Review by Zynton — COMMENTED** [review]
@awa-odoo It sure is head crunching once you try to do everything _as if there was no iframe in the document_ :eyes: Thanks for making the effort of being this thorough nonetheless :-)

Anyway, I responded to your comments and pushed my changes. The action thing is a bit of a puzzle though.

File: addons/mass_mailing/views/mailing_mailing_views.xml [hunk 33]
```diff
@@ -459,6 +534,12 @@
             <field name="name">Mailings</field>
             <field name="res_model">mailing.mailing</field>
             <field name="view_mode">tree,kanban,form,calendar,graph</field>
+            <field name="view_ids" eval="[(5, 0, 0),
+                (0, 0, {'view_mode': 'tree'}),
+                (0, 0, {'view_mode': 'kanban'}),
+                (0, 0, {'view_mode': 'form', 'view_id': ref('mailing_mailing_view_full_width_form')}),
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mass_mailing/views/mailing_mailing_views.xml]: It is indeed a bit verbose but easier for future changes / customizations / understanding views ...

File: addons/mass_mailing/static/src/js/mass_mailing_full_width_form.js [hunk 34]
```diff
@@ -1,10 +1,150 @@
...
+        }
+    },
+    /**
+     * Return true if the mailing editor is in full screen mode, false
+     * otherwise.
+     *
+     * @private
+     * @returns {boolean}
+     */
+   _isFullScreen() {
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_full_width_form.js]: Linting: indentation issue (seems fixed in next commit ? probably a rebase to do to avoid commit fixups)

File: addons/mass_mailing/views/mailing_mailing_views.xml [hunk 35]
```diff
@@ -315,6 +316,80 @@
             </field>
         </record>
 
+        <!-- Inherited form view for mass mailing's form specifically -->
+        <record model="ir.ui.view" id="mailing_mailing_view_full_width_form">
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/views/mailing_mailing_views.xml:320]: mailing_mailing_view_form_full_width ? just to follow guidelines (view_form indicates a form view, then a suffix for details)

File: addons/mass_mailing/views/mailing_mailing_views.xml [hunk 36]
```diff
@@ -315,6 +316,80 @@
...
+            <field name="name">mailing.mailing.form_inherit</field>
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/views/mailing_mailing_views.xml:321]: Name should globally be the same as xml id (with dots instead of underscores)

File: addons/mass_mailing/views/mailing_mailing_views.xml [hunk 37]
```diff
@@ -315,6 +316,80 @@
...
+            <field name="inherit_id" ref="mass_mailing.view_mail_mass_mailing_form"/>
+            <field name="mode">primary</field>
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/views/mailing_mailing_views.xml:323]: Well we should probably force its priority to ensure it is not used as default view, as it is used only in email marketing context (otherwise fallback on standard view, unless I misunderstood something)

File: addons/mass_mailing/views/mailing_mailing_views.xml [hunk 38]
```diff
@@ -315,6 +316,80 @@
...
+            <field name="model">mailing.mailing</field>
+            <field name="arch" type="xml">
+                <xpath expr="//form" position="attributes">
+                    <attribute name="js_class">mass_mailing_full_width_form</attribute>
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/views/mailing_mailing_views.xml:327]: Would maybe use something like mailing_mailing_view_form_full_width to standardize naming a bit (and use the same as xml id, so that all naming is coherent and grep is easy)

File: addons/web/static/src/legacy/js/core/dom.js [hunk 39]
```diff
@@ -545,11 +545,14 @@ var dom = {
...
      *      ignored too)
+     * @param {JQuery} [options.$scrollable] the $element to scroll
      * @return {Promise}
      */
     scrollTo(el, options = {}) {
         const $el = $(el);
-        const $scrollable = $el.parent().closestScrollable();
+        const $scrollable = options.$scrollable || $el.parent().closestScrollable();
+        const isInOneDocument = $scrollable[0].ownerDocument === $el[0].ownerDocument;
+        const $iframe = !isInOneDocument && $scrollable.find('iframe').filter((i, node) => $(node).contents().has($el));
```
Comment (tde-banana-odoo) [inline addons/web/static/src/legacy/js/core/dom.js:555]: Those lines probably could gain some explanations ^^ don't understand them. I suppose it is quite clear for you, but for external people, well :D ...

File: addons/web/static/src/legacy/js/core/dom.js [hunk 40]
```diff
@@ -545,11 +545,14 @@ var dom = {
      * @param {number} [options.forcedOffset]
      *      offset used instead of the automatic one (extraOffset will be
      *      ignored too)
+     * @param {JQuery} [options.$scrollable] the $element to scroll
      * @return {Promise}
      */
     scrollTo(el, options = {}) {
         const $el = $(el);
-        const $scrollable = $el.parent().closestScrollable();
+        const $scrollable = options.$scrollable || $el.parent().closestScrollable();
```
Comment (tde-banana-odoo) [inline addons/web/static/src/legacy/js/core/dom.js:553]: A bit strange to have a scroll To taking an element as parameter, but target is coming from options ... I understand the global issue, but seems a bit coding by side-effects... I don't have any better proposal as this is not my main scope, just wondering a bit.

**Review by tde-banana-odoo — COMMENTED** [review]
Thanks for the PR, looks indeed better contained than first PR :) . I have some additional comments and questions, but globally seems good :) .

File: addons/mass_mailing/views/mailing_mailing_views.xml (hunk 36 above)
Comment (Zinston) [reply addons/mass_mailing/views/mailing_mailing_views.xml:321]: So `mailing.mailing.view.form.full_width` ?

Comment (tde-banana-odoo) [reply addons/mass_mailing/views/mailing_mailing_views.xml:321]: more ``mailing.mailing.view.form.full.width`` (to avoid mix of dots and underscores). As name is not really used on views this is a standard way of filling it.

File: addons/web/static/src/legacy/js/core/dom.js (hunk 39 above)
Comment (Zinston) [reply addons/web/static/src/legacy/js/core/dom.js:558]: Haha sure ok :p

```js
// If $scrollable and $el are not in the same document, we can safely
// assume $el is in an $iframe. We retrieve it by filtering the list of
// iframes in $scrollable to keep only the one that contains $el.
```

File: addons/mass_mailing/views/mailing_mailing_views.xml (hunk 37 above)
Comment (Zinston) [reply addons/mass_mailing/views/mailing_mailing_views.xml:323]: I'm not sure what you mean. Do you mean I should add like `<field name="priority">2</field>` on this one and `<field name="priority">1</field>` on the other ?

**Review by Zynton — COMMENTED** [review]
@tde-banana-odoo Thanks for the review, glad we're moving forward! I applied what I could but had a couple of questions about your remarked so I put them here.

**Comment by tde-banana-odoo** [comment]
@robodoo r+

**Comment by tde-banana-odoo** [comment]
@Zinston LGTM, thanks for answering and updating :) . I think we can merge and move forward !

**Comment by tde-banana-odoo** [comment]
@robodoo rebase-merge
